qiskit/quantum_info/synthesis/xx_decompose/polytopes.py

```diff
@@ -0,0 +1,263 @@
+which describes those two-qubit programs accessible to a given sequence of XX-type interactions.
+"""
+
+from copy import copy
+from dataclasses import dataclass, field
+from itertools import combinations
+from typing import List
+
+import numpy as np
+
+from qiskit.exceptions import QiskitError
+
+from .utilities import EPSILON
+
+
+@dataclass
+class ConvexPolytopeData:
+    """
+    The raw data underlying a ConvexPolytope.  Describes a single convex
+    polytope, specified by families of `inequalities` and `equalities`, each
+    entry of which respectively corresponds to
+
+        inequalities[j][0] + sum_i inequalities[j][i] * xi >= 0
+
+    and
+
+        equalities[j][0] + sum_i equalities[j][i] * xi == 0.
+    """
+
+    inequalities: List[List[int]]
+    equalities: List[List[int]] = field(default_factory=list)
+    name: str = ""
+
+
+@dataclass
+class PolytopeData:
+    """
+    The raw data of a union of convex polytopes.
+    """
+
+    convex_subpolytopes: List[ConvexPolytopeData]
+
+
+def polytope_has_element(polytope, point):
+    """
+    Tests whether `polytope` contains `point.
+    """
+    return all(
+        -EPSILON <= inequality[0] + sum(x * y for x, y in zip(point, inequality[1:]))
+        for inequality in polytope.inequalities
+    ) and all(
+        abs(equality[0] + sum(x * y for x, y in zip(point, equality[1:]))) <= EPSILON
+        for equality in polytope.equalities
+    )
+
+
+def manual_get_vertex(polytope):
+    """
+    Returns a single "random" vertex from `polytope`.
+
+    NOTE: Not actually random, will return the same vertex on each call --- but it has no preference
+          for which vertex it returns, and it makes no guarantee to the caller beyond consistency.
+    """
+    if isinstance(polytope, PolytopeData):
+        paragraphs = copy(polytope.convex_subpolytopes)
+    elif isinstance(polytope, ConvexPolytopeData):
+        paragraphs = [polytope]
+    else:
+        raise TypeError(f"{type(polytope)} is not polytope-like.")
+
+    for convex_subpolytope in paragraphs:
+        sentences = convex_subpolytope.inequalities + convex_subpolytope.equalities
+        if len(sentences) == 0:
+            continue
+        dimension = len(sentences[0]) - 1
+        for inequalities in combinations(sentences, dimension):
+            matrix = np.array([x[1:] for x in inequalities])
+            b = np.array([x[0] for x in inequalities])
+            try:
+                vertex = np.linalg.inv(-matrix) @ b
+                if polytope_has_element(convex_subpolytope, vertex):
+                    return vertex
+            except np.linalg.LinAlgError:
+                pass
+
+    raise QiskitError(f"Polytope has no feasible solutions:\n{polytope}")
+
+
+@dataclass
+class XXPolytope:
+    """
+    Describes those two-qubit programs accessible to a given sequence of XX-type interactions.
+
+    Strengths are normalized so that CX corresponds to pi / 4.
```

Comment:
i guess this is different from the Qiskit convention, so maybe worth an explicit note